### PR TITLE
`Iris`: Enhance student support with proactive assistance

### PR DESCRIPTION
**Pull request by kaancayli**
# ⚠️  Contains Database Migration - Deploy on TS3
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Implements the feature proposal described in #8800 

### Description

- Added a new proactivity feature to Iris
- Added PyrisEventService that dispatches the detected events to their handler methods
- Added eventVariant to pipeline execution parameters
- Defined 3 event types Iris can react to (jol, build_failed, progress_stalled)
- Build failed event gets triggered when the a submission fails to build
- Progress stalled event is triggered, when the student's score is not increasing after at least 3 subsequent submissions
- Added proactivity settings in Iris chat settings for the new proactivity feature
- Added event payloads to pipeline execution DTOs

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise with Iris Enabled
- Iris terms and conditions are already accepted
- `Monitor student submission build failures` under Iris Chat Settings Enabled
- `Monitor exercise performance progress` under Iris Chat Settings Enabled
-  ls1intum/Pyris#154 merged or deployed on pyris-test

1. Log in to Artemis
2. Navigate to the programming exercise with Iris enabled
3. Make sure the proactive event options under Iris Chat Settings are enabled in all levels:
   - Global
   - Course
   - Exercise

4. Testing events:
   1. Build failed event:
      - Make a programming exercise submission with a failing build
      - Verify that Iris reacted and sent a proactive message related to the error
   
   2. Progress stalled event:
      - Make 3 consequent programming exercise submissions
      - Make sure the scores after each submission are not increasing (e.g., you can submit the initial repository 3 times)
      - Verify that Iris reacted and sent a proactive message related to the error

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| iris-sub-settings.model.ts | 100% | ✅ |
| exercise-chatbot-button.component.ts | 87.3% | ✅  |
| iris-chat.service.ts | 89.04% | ✅  |
| iris-common-sub-settings-update.component.ts | 88.88% | ✅  |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| IrisCourseSettings.java | 100% | ✅ |
| IrisExerciseSettings.java | 94% | ✅  |
| IrisGlobalSettings.java | 100% | ✅ |
| IrisSettings.java | 100% | ✅ |
| IrisSubSettings.java | 100% | ✅ |
| IrisSubSettingsType.java | 100% | ✅ |
| IrisCombinedSettingsDTO.java | 100% | ✅ |
| IrisSettingsRepository.java | 100% | ✅ |
| IrisCompetencyGenerationService.java | 82% | ✅  |
| PyrisConnectorService.java | 79% | ✅  |
| PyrisEventService.java | 100% | ✅ |
| PyrisPipelineService.java | 80% | ✅ |
| PyrisEventDTO.java | 100% | ✅ |
| PyrisCourseChatPipelineExecutionDTO.java | 100% | ✅ |
| NewResultEvent.java | 100% | ✅ |
| PyrisEvent.java | 100% | ✅ |
| IrisCourseChatSessionService.java | 68% | ✅  |
| IrisExerciseChatSessionService.java | 91% | ✅  |
| IrisTextExerciseChatSessionService.java | 86% | ✅  |
| IrisSettingsService.java | 91% | ✅  |
| IrisSubSettingsService.java | 91% | ✅  |
| IrisExerciseChatSessionResource.java | 91% | ✅  |

#### Proactivity Settings
Screenshot 2024-11-21 at 23 41 00
Screenshot 2024-11-21 at 23 31 23

#### Proactive responses

ScreenRecording2024-11-21at23 37 59-ezgif com-video-to-gif-converter

## Release Notes

- **New Features**
  - Introduced new proactive event management settings for chat, allowing users to enable or disable events based on their preferences.
  - Added support for new event types, including `BUILD_FAILED` and `PROGRESS_STALLED`, enhancing the system's responsiveness to user interactions.
  - Enhanced the chat interface with a message bubble feature that previews new messages.
  - Added new methods for managing submissions and event notifications, improving integration with the Pyris event system.
  - Implemented new exception classes for better error handling related to event processing.
  - Added a new method for retrieving submissions with results, improving data access.
  - Introduced a new `PyrisEventDTO` for encapsulating event data along with its type.
  - Added functionality for managing proactive settings directly within the UI for chat sub-settings.

- **Bug Fixes**
  - Improved error handling for unsupported events and refined logging for better traceability.

- **Documentation**
  - Updated documentation to reflect the new event types and proactive settings.

- **Tests**
  - Expanded test coverage for new event management features and ensured robust handling of edge cases in event processing.
  - Added integration tests for the Pyris event system to validate behavior under various scenarios.

### CODE REVIEW

**Review by alexjoham — APPROVED** [review]
The code looks good!
I tested everything locally and on TS3, and it works as it should; token tracking is also already implemented and working!

**Review by yassinsws — DISMISSED** [review]
Code approved, made some comments but only nitpicks. Good job Kaan
Tested on ts3 works as expected !

**Review by sebastianloose — APPROVED** [review]
The code looks good to me! I tested on TS3 and everything worked as expected

File: src/main/java/de/tum/cit/aet/artemis/iris/domain/settings/event/IrisEventType.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+    private final String name;
+
+    IrisEventType(String name) {
+        this.name = name;
+    }
+
+    public String getName() {
+        return name;
+    }
```
[the remark this answers is not in the document]
Comment (kaancayli) [reply src/main/java/de/tum/cit/aet/artemis/iris/domain/settings/event/IrisEventType.java:18]: That's due to name of the events defined in Pyris. In Pyris the event flags are defined in lower case, and the  architectural tests enforces that `.toString()` and `.name()` of an enum return the same value.

[the remark this answers is not in the document]
Comment (kaancayli) [reply src/main/java/de/tum/cit/aet/artemis/iris/domain/settings/event/IrisEventType.java]: I have applied your suggestion regardless 😄

File: src/main/webapp/app/iris/exercise-chatbot/exercise-chatbot-button.component.ts [hunk 2]
```diff
@@ -38,7 +71,7 @@ export class IrisExerciseChatbotButtonComponent implements OnInit, OnDestroy {
     ) {}
 
     ngOnInit() {
-        // Subscribes to route params and gets the exerciseId from the route
+        // Subscribes to route params and gets the exerciseId from the router
```
Comment (SimonEntholzer) [inline src/main/webapp/app/iris/exercise-chatbot/exercise-chatbot-button.component.ts:74]: I think the comment was correct

File: src/main/webapp/app/iris/exercise-chatbot/exercise-chatbot-button.component.ts [hunk 3]
```diff
@@ -1,31 +1,64 @@
...
     private numNewMessagesSubscription: Subscription;
     private paramsSubscription: Subscription;
+    private latestIrisMessageSubscription: Subscription;
 
     // Icons
     faCircle = faCircle;
     faChevronDown = faChevronDown;
+    faAngleDoubleDown = faAngleDoubleDown;
+
+    @ViewChild('chatBubble') chatBubble: ElementRef;
```
Comment (SimonEntholzer) [inline src/main/webapp/app/iris/exercise-chatbot/exercise-chatbot-button.component.ts:62]: Can you adapt the file to use the new client guidelines? With inject instead  of the constructor, and using signals?

File: src/main/webapp/app/iris/settings/iris-settings-update/iris-common-sub-settings-update/iris-common-sub-settings-update.component.html [hunk 4]
```diff
@@ -43,6 +43,40 @@ <h4 class="form-label mt-3" jhiTranslate="artemisApp.iris.settings.subSettings.e
...
+                        [disabled]="!isAdmin && !parentSubSettings"
+                        [ngModel]="!subSettings?.disabledProactiveEvents?.includes(event)"
+                        (ngModelChange)="onEventToggleChange(event)"
+                    />
+                    <label class="form-check-label" for="{{ event }}" [jhiTranslate]="eventTranslationKeys[event]"></label>
+                    @if (
+                        this.parentSubSettings &&
+                        !subSettings?.disabledProactiveEvents?.includes(event) &&
+                        (this.parentSubSettings.disabledProactiveEvents?.includes(event) || !this.parentSubSettings.enabled)
+                    ) {
```
Comment (SimonEntholzer) [inline src/main/webapp/app/iris/settings/iris-settings-update/iris-common-sub-settings-update/iris-common-sub-settings-update.component.html:68]: would it be possible to move this check out of the html? Having several function calls in here isn't best practice, as they are often re-evaluated

File: src/main/webapp/app/iris/settings/iris-settings-update/iris-common-sub-settings-update/iris-common-sub-settings-update.component.ts [hunk 5]
```diff
@@ -52,11 +54,18 @@ export class IrisCommonSubSettingsUpdateComponent implements OnInit, OnChanges {
...
     protected readonly IrisSubSettingsType = IrisSubSettingsType;
 
+    protected readonly eventTranslationKeys = {
+        [IrisEventType.BUILD_FAILED]: 'artemisApp.iris.settings.subSettings.proactivityBuildFailedEventEnabled.label',
+        [IrisEventType.PROGRESS_STALLED]: 'artemisApp.iris.settings.subSettings.proactivityProgressStalledEventEnabled.label',
+    };
+
     constructor(
         accountService: AccountService,
         private irisSettingsService: IrisSettingsService,
```
Comment (SimonEntholzer) [inline src/main/webapp/app/iris/settings/iris-settings-update/iris-common-sub-settings-update/iris-common-sub-settings-update.component.ts:71]: Could you adapt this file to the new client guidelines?

File: src/main/webapp/i18n/en/iris.json [hunk 6]
```diff
@@ -32,6 +32,19 @@
...
+                        "tooltip": "When enabled, Iris will proactively send help messages if a submission build fails."
+                    },
+                    "proactivityProgressStalledEventEnabled": {
+                        "label": "Monitor exercise performance progress",
+                        "tooltip": "When enabled, Iris will proactively send help messages if the student's exercise performance does not improve."
+                    },
+                    "proactivitySettings": {
+                        "title": "Proactivity Settings",
+                        "tooltip": "When any of options listed below are activated, Iris will proactively send help messages to students based on the selected conditions.",
+                        "parentDisabled": "Iris won't send proactive messages for this option because either the same option or the Iris is disabled in the parent settings."
```
Comment (SimonEntholzer) [inline src/main/webapp/i18n/en/iris.json:46]: This sounds really clunky imo. Maybe rewrite it something along the lines of:

"Iris won't send proactive messages for this option because either option itself or Iris are disabled in the parent settings"

Not 100% sure though tbh what you are trying to say here. (Same for the german version)

File: src/test/java/de/tum/cit/aet/artemis/iris/PyrisEventSystemTest.java [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+import de.tum.cit.aet.artemis.iris.service.session.IrisExerciseChatSessionService;
+import de.tum.cit.aet.artemis.programming.domain.ProgrammingExercise;
+import de.tum.cit.aet.artemis.programming.domain.ProgrammingExerciseStudentParticipation;
+import de.tum.cit.aet.artemis.programming.domain.ProgrammingSubmission;
+import de.tum.cit.aet.artemis.programming.domain.ProjectType;
+import de.tum.cit.aet.artemis.programming.domain.SolutionProgrammingExerciseParticipation;
+import de.tum.cit.aet.artemis.programming.domain.TemplateProgrammingExerciseParticipation;
+import de.tum.cit.aet.artemis.programming.util.ProgrammingExerciseUtilService;
+
+class PyrisEventSystemTest extends AbstractIrisIntegrationTest {
```
Comment (SimonEntholzer) [inline src/test/java/de/tum/cit/aet/artemis/iris/PyrisEventSystemTest.java:58]: Would it make sense to rename it to PyrisEventSystemIntegrationTest, since it is an integration test?

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
Nice changes, I added some small improvement suggestions

File: src/main/java/de/tum/cit/aet/artemis/iris/service/session/IrisCourseChatSessionService.java [hunk 8]
```diff
@@ -134,13 +133,13 @@ protected void setLLMTokenUsageParameters(LLMTokenUsageService.LLMTokenUsageBuil
      */
     public void onJudgementOfLearningSet(CompetencyJol competencyJol) {
         var course = competencyJol.getCompetency().getCourse();
-        if (!irisSettingsService.isEnabledFor(IrisSubSettingsType.CHAT, course)) {
-            return;
-        }
+
+        // TODO: Add setting to activate/deactivate course chat for JOLs
+
```
[the remark this answers is not in the document]
Comment (kaancayli) [reply src/main/java/de/tum/cit/aet/artemis/iris/service/session/IrisCourseChatSessionService.java:138]: Yes, it will come in a follow-up PR. Currently there is an open PR for adding course chat settings, and the event is related to course chat settings.

See #9866

File: src/main/java/de/tum/cit/aet/artemis/iris/service/session/IrisExerciseChatSessionService.java [hunk 9]
```diff
@@ -57,11 +72,19 @@ public class IrisExerciseChatSessionService extends AbstractIrisChatSessionServi
 
     private final ProgrammingExerciseRepository programmingExerciseRepository;
 
+    private final IrisExerciseChatSessionRepository irisExerciseChatSessionRepository;
+
+    private final SubmissionRepository submissionRepository;
+
+    private final double SUCCESS_THRESHOLD = 100.0; // TODO: Retrieve configuration from Iris settings
+
+    private final int INTERVAL_SIZE = 3; // TODO: Retrieve configuration from Iris settings
```
[the remark this answers is not in the document]
Comment (kaancayli) [reply src/main/java/de/tum/cit/aet/artemis/iris/service/session/IrisExerciseChatSessionService.java:81]: Yes, currently we are in discussion how we can best integrate the event specific settings into the existing Iris settings structure. I have made of them a local variable though.

File: src/main/webapp/app/iris/exercise-chatbot/exercise-chatbot-button.component.ts (hunk 3 above)
Comment (kaancayli) [reply src/main/webapp/app/iris/exercise-chatbot/exercise-chatbot-button.component.ts:62]: There is currently a separate PR for migrating Iris components according to latest coding guidelines for angular.

Please see #9615

File: src/main/java/de/tum/cit/aet/artemis/iris/service/session/IrisExerciseChatSessionService.java [hunk 10]
```diff
@@ -152,9 +261,127 @@ private Optional<ProgrammingSubmission> getLatestSubmissionIfExists(ProgrammingE
...
+
+    /**
+     * Checks if the student needs intervention based on their recent score trajectory.
+     *
+     * @param scores       The list of all scores for the student.
+     * @param intervalSize The number of recent submissions to consider.
+     * @return true if intervention is needed, false otherwise.
+     */
+    private boolean needsIntervention(List<Double> scores, int intervalSize) {
+        if (scores.size() < intervalSize) {
```
[the remark this answers is not in the document]
Comment (kaancayli) [reply src/main/java/de/tum/cit/aet/artemis/iris/service/session/IrisExerciseChatSessionService.java:288]: Done

**Review by coolchock — APPROVED** [review]
tested on ts3, iris messages are displayed both for 3 subsequent submissions with the same score and build failures

File: src/main/webapp/i18n/en/iris.json (hunk 6 above)
Comment (kaancayli) [reply src/main/webapp/i18n/en/iris.json:46]: I have adjusted the translations according to your suggestion

File: src/main/webapp/app/iris/settings/iris-settings-update/iris-common-sub-settings-update/iris-common-sub-settings-update.component.ts (hunk 5 above)
Comment (kaancayli) [reply src/main/webapp/app/iris/settings/iris-settings-update/iris-common-sub-settings-update/iris-common-sub-settings-update.component.ts:71]: There is currently a separate PR for migrating Iris components according to latest coding guidelines for angular.

Please see #9615

File: src/main/webapp/app/iris/settings/iris-settings-update/iris-common-sub-settings-update/iris-common-sub-settings-update.component.html (hunk 4 above)
Comment (kaancayli) [reply src/main/webapp/app/iris/settings/iris-settings-update/iris-common-sub-settings-update/iris-common-sub-settings-update.component.html:68]: I have moved the logic into the `ngOnChanges` life-cycle hook.

File: src/main/java/de/tum/cit/aet/artemis/iris/service/session/IrisExerciseChatSessionService.java [hunk 11]
```diff
@@ -140,7 +175,81 @@ public void requestAndHandleResponse(IrisExerciseChatSession session) {
...
+        var participation = result.getSubmission().getParticipation();
+        if (!(participation instanceof ProgrammingExerciseStudentParticipation studentParticipation)) {
+            return;
+        }
+
+        var exercise = validateExercise(participation.getExercise());
+
+        irisSettingsService.isActivatedForElseThrow(IrisEventType.PROGRESS_STALLED, exercise);
+
+        var recentSubmissions = submissionRepository.findAllWithResultsAndAssessorByParticipationIdOrderBySubmissionDateAsc(studentParticipation.getId());
```
[the remark this answers is not in the document]
Comment (kaancayli) [reply src/main/java/de/tum/cit/aet/artemis/iris/service/session/IrisExerciseChatSessionService.java:224]: Done

File: src/main/webapp/i18n/de/iris.json [hunk 12]
```diff
@@ -32,6 +32,19 @@
...
+                        "tooltip": "Wenn aktiviert, wird Iris proaktiv Hilfe-Nachrichten senden, wenn der Build einer Abbgabe fehlschlägt."
+                    },
+                    "proactivityProgressStalledEventEnabled": {
+                        "label": "Übungsleistung überwachen",
+                        "tooltip": "Wenn aktiviert, wird Iris proaktiv Hilfe-Nachrichten senden, wenn sich die Übungsleistung des Studenten nicht verbessert."
+                    },
+                    "proactivitySettings": {
+                        "title": "Proaktivitätseinstellungen",
+                        "tooltip": "Aktiviere die untenstehenden Optionen, damit Iris proaktiv Studierende kontaktiert, wenn bestimmte Bedingungen erfüllt sind.",
+                        "parentDisabled": "Iris sendet keine proaktiven Nachrichten für diese Option, da entweder die Option selbst oder Iris in den aktuellen oder übergeordneten Einstellungen deaktiviert sind."
```
Comment (SimonEntholzer) [inline src/main/webapp/i18n/de/iris.json:46]:
```suggestion
                        "parentDisabled": "Iris sendet keine proaktiven Nachrichten für diese Option, da entweder die Option selbst oder Iris in den aktuellen oder übergeordneten Einstellungen deaktiviert ist."
```

File: src/main/webapp/i18n/en/iris.json [hunk 13]
```diff
@@ -32,6 +32,19 @@
...
+                        "tooltip": "When enabled, Iris will proactively send help messages if a submission build fails."
+                    },
+                    "proactivityProgressStalledEventEnabled": {
+                        "label": "Monitor exercise performance progress",
+                        "tooltip": "When enabled, Iris will proactively send help messages if the student's exercise performance does not improve."
+                    },
+                    "proactivitySettings": {
+                        "title": "Proactivity Settings",
+                        "tooltip": "Enable options below to allow Iris to proactively reach out to students when specific conditions are met.",
+                        "parentDisabled": "Iris won't send proactive messages for this option because either option itself or Iris are disabled in the current or parent settings."
```
Comment (SimonEntholzer) [inline src/main/webapp/i18n/en/iris.json:46]: I think that's a lot more understandable now, thx :D only one small change request left:

```suggestion
                        "parentDisabled": "Iris won't send proactive messages for this option because either the option itself or Iris are disabled in the current or parent settings."
```

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
thx for the changes :D only one small grammar fix left

File: src/main/webapp/i18n/en/iris.json (hunk 13 above)
Comment (kaancayli) [reply src/main/webapp/i18n/en/iris.json]: Addressed the change request

File: src/main/webapp/i18n/de/iris.json (hunk 12 above)
Comment (kaancayli) [reply src/main/webapp/i18n/de/iris.json]: Addressed the change request

**Review by yassinsws — DISMISSED** [review]
reapprove !

**Review by SimonEntholzer — DISMISSED** [review]
Code lgtm now

**Review by alexjoham — APPROVED** [review]
Reapprove 👍🏻

**Review by sebastianloose — DISMISSED** [review]
LGTM reapprove!

**Review by coolchock — APPROVED** [review]
re-approve

**Review by HawKhiem — DISMISSED** [review]
Tested on TS3. After a build fail, Iris sent a message pointing out the error.

Screenshot 2024-12-02 120840

After 3 successive submissions with 0% (inluding the first build fail), Iris sent a message giving me hint to implement the problem
Screenshot 2024-12-02 121047

**Review by alexjoham — APPROVED** [review]
Code sill looks good to me! Reapprove!

**Review by isabellagessl — DISMISSED** [review]
Code looks good to me :)

**Review by sebastianloose — DISMISSED** [review]
Reapprove 👍

**Review by yassinsws — DISMISSED** [review]
reapprove

**Review by HawKhiem — APPROVED** [review]
Reapprove

**Review by coolchock — APPROVED** [review]
re-approve

**Review by sebastianloose — APPROVED** [review]
Reapprove

**Review by SimonEntholzer — APPROVED** [review]
Re-approve code

**Review by alexjoham — DISMISSED** [review]
Reapprove!

**Review by coolchock — DISMISSED** [review]
re-approve

**Review by isabellagessl — APPROVED** [review]
Reapprove :)

**Review by sebastianloose — APPROVED** [review]
Reapprove 👍

**Review by yassinsws — DISMISSED** [review]
reapprove

**Review by sebastianloose — APPROVED** [review]
Reapprove

**Review by isabellagessl — APPROVED** [review]
Tested on Test Server 3 look good :)
